### ISSUE TITLE
STRTRIM() is around 20 times slower in GDL compared to IDL

### ISSUE THREAD

**Issue by eloirozier** [gnudatalanguage/gdl#1045]
On my Ubuntu laptop:

```
GDL> tic & for i=0, 100000 do a=STRTRIM('1.25',2) & toc
% Time elapsed :       0.42429996 seconds.
```

```
IDL> tic & for i=0, 100000 do a=STRTRIM('1.25',2) & toc
% Time elapsed: 0.023247004 seconds.
```

It is critical when intensely calling the READCOL procedure (from `idlastro` ), GDL is then VERY slow compared to IDL
I came upon this while making test runs with PSM (Planck Sky Model https://apc.u-paris.fr/~delabrou/PSM/psm.html)

**Comment by eloirozier** [comment]
functions STRTRIM() and STRCOMPRESS() are also 10 times slower (than IDL)

**Comment by GillesDuvert** [comment]
Yes, but,  and this is not particular for STRTRIM, you get the same time if the strtrim() function returns immediately after being called (just add  `return new DLongGDL(0);` at the beginning of the function) .
In other words, any GDL command takes 20 times more time to be called than IDL's.
There is plenty of room for improvement, but not in the STRTRIM function itself.
I suspect the C++ object creations involved in all the hidden process.

**Comment by GillesDuvert** [comment]
Generally speaking, if a procedure is slow because of a loop, it is because the vectorial aspect of IDL(GDL) has been neglected. Specifically READCOL claims to be slow and not to be the thing to use if the format of the (ascii-encoded!) data is known. Which is perhaps the case for PSM?

**Comment by alaingdl** [comment]
I made tests in another way and results are good concerning STRMID()

```
a=read_ascii_basic('CMakeLists.txt')  ;  text file with 1216 lines today
ab=[a,a,a,a,a]
tic & for i=0, 100 do b=STRTRIM(ab,2) & toc
tic & for i=0, 100 do b=STRTRIM(ab,1) & toc
tic & for i=0, 100 do b=STRTRIM(ab,0) & toc
```
IDL (8.4) and GDL behaved very very similar in time (after few re-runs) (all runs are close to 0.12 s on my laptop)

**Comment by GillesDuvert** [comment]
Another (stranger) example:
```
GDL> tic & for i=0, 100000 do wait,0.00001 & toc
% Time elapsed :        1.0504420 seconds.
IDL> tic & for i=0, 100000 do wait,0.00001 & toc 
% Time elapsed: 6.5285769 seconds.

```
(No comment on the performance of IDL in this case :smile: ) 

but:
```
IDL> tic & for i=0, 100000 do a=0 & toc          
% Time elapsed: 0.0073997974 seconds.
GDL>  tic & for i=0, 100000 do a=0 & toc 
% Time elapsed :       0.12551284 seconds.

```
so the 1.05 second for GDL in the first case means that 100000 loops take 0.05 seconds, and the second example show that 100000 creations of 'a' and assignment to 0 takes 0.12551284-0.0504420=0.07 seconds, 10x IDL's.

**Comment by eloirozier** [comment]
@GillesDuvert  This is enlightening. However if I do as you tell, (return at the very beginning of strtrim()), it is then two times faster. So there might still be room for improvement, but at least now i know why i will not get any faster than IDL.

**Comment by alaingdl** [comment]
Another point for READCOL, recent versions (since 0.9.9) of FILE_LINES() are about 5 to 10 times slower than the version in Vanilla 0.9.8 (on a file with 100 003 ascii lines, tested on two computers)

```
IDL>  tic & print, file_lines('~/pointsD.txt') & toc
                100003
% Time elapsed: 0.051157951 seconds.

GDL 0.9.8 Vanilla> tic & print, file_lines('~/pointsD.txt') & toc
% Compiled module: FILE_LINES.
                100003
% Time elapsed :      0.043983221 seconds.

GDL git> tic & print, file_lines('~/pointsD.txt') & toc
      100003
% Time elapsed :       0.18068409 seconds.
```

Please note that in GDL 0.9.8 we do use an *.pro function ... If I used the old pro file (renaming it into file_lines2) I have : 
```
GDL> tic & print, file_lines2('~/pointsD.txt') & toc
                100003
% Time elapsed :      0.044064999 seconds.
GDL> tic & print, file_lines('~/pointsD.txt') & toc
      100003
% Time elapsed :       0.31277704 seconds.
```

CQFD (as we say in French :smile:)

**Comment by GillesDuvert** [comment]
yes, but:
GDL  0.9.8 (procedure):
```
`GDL> tic & for i=1,100 do a=file_lines('~/.bashrc') & toc
% Time elapsed :       0.75447607 seconds.
```
GDL 1.0.0 (C++ code)
```
GDL> tic & for i=1,100 do a=file_lines('~/.bashrc') & toc
% Time elapsed :      0.026001215 seconds.

```
IDL does this in 0.0056s and is still 5 times better than GDL.

As the 'file_lines.pro' procedure used to spawn the linux 'wc' command, it is disqualified for a general useage anyway.

**Comment by alaingdl** [comment]
We (@eloirozier , @bouloukaouze , @alaingdl ) do not forgot that and we are back.

* We have to fix the new FILE_LINES()
* We started an audit on some STRING related functions, and we do have some decrepencies but not espécially in STRTRIM(). More details to come
